chain/checkpointing/sub.go

```diff
@@ -0,0 +1,506 @@
+package checkpointing
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"os"
+	"sort"
+	"strconv"
+
+	"github.com/BurntSushi/toml"
+	"github.com/Zondax/multi-party-sig/pkg/math/curve"
+	"github.com/Zondax/multi-party-sig/pkg/party"
+	"github.com/Zondax/multi-party-sig/pkg/protocol"
+	"github.com/Zondax/multi-party-sig/pkg/taproot"
+	"github.com/Zondax/multi-party-sig/protocols/frost"
+	"github.com/Zondax/multi-party-sig/protocols/frost/keygen"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/blockstore"
+	"github.com/filecoin-project/lotus/chain/consensus/actors/mpower"
+	"github.com/filecoin-project/lotus/chain/events"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/node/impl"
+	"github.com/filecoin-project/lotus/node/modules/helpers"
+	cbor "github.com/ipfs/go-ipld-cbor"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p-core/host"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"go.uber.org/fx"
+)
+
+var log = logging.Logger("checkpointing")
+
+type CheckpointingSub struct {
+	host   host.Host
+	pubsub *pubsub.PubSub
+	// Topic for keygen
+	topic *pubsub.Topic
+	// Sub for keygen
+	sub *pubsub.Subscription
+	// This is the API for the fullNode in the root chain.
+	api *impl.FullNodeAPI
+	// Listener for events of the root chain.
+	events *events.Events
+	// Generated public key
+	pubkey []byte
+	// taproot config
+	config *keygen.TaprootConfig
+	// new config generated
+	newconfig *keygen.TaprootConfig
+	// Initiated
+	init bool
+	// Previous tx
+	ptxid string
+	// Tweaked value
+	tweakedValue []byte
+}
+
+func NewCheckpointSub(
+	mctx helpers.MetricsCtx,
+	lc fx.Lifecycle,
+	host host.Host,
+	pubsub *pubsub.PubSub,
+	api impl.FullNodeAPI,
+) (*CheckpointingSub, error) {
+
+	ctx := helpers.LifecycleCtx(mctx, lc)
+	// Starting shardSub to listen to events in the root chain.
+	e, err := events.NewEvents(ctx, &api)
+	if err != nil {
+		return nil, err
+	}
+
+	var config *keygen.TaprootConfig
+	// Load configTaproot
+	if _, err := os.Stat(os.Getenv("EUDICO_PATH") + "/share.toml"); errors.Is(err, os.ErrNotExist) {
+		// path/to/whatever does not exist
+		fmt.Println("No share file saved")
```

Comment:
Sanity-check: We do nothing and return an empty config if no share is stored here? Don't we need to return an error or do somethign to avoid the config from being `nil`?

If this is the case we could probably change this if statement to make it more legible and do something like:
```go
if (dir doesn't exist) {
   // Generate config
}
```

Reply:
Oh, I see, if there is no config you just start an empty ShardingSub to be populated in the future. Then you probably don't need the `else` and a single `if' as proposed can do the trick.